istioctl/cmd/generate_cert_key.go

```diff
@@ -0,0 +1,225 @@
+	generateKeyAndCertCmd.PersistentFlags().StringVar(&org, "organization", "Juju org", "Organization for the cert.")
+	generateKeyAndCertCmd.PersistentFlags().StringVar(&outCert, "out-cert", "cert-chain.pem", "Output certificate file.")
+	generateKeyAndCertCmd.PersistentFlags().StringVar(&outPriv, "out-priv", "key.pem", "Output private key file.")
+	generateKeyAndCertCmd.PersistentFlags().IntVar(&keySize, "key-size", 2048, "Size of the generated private key")
+	generateKeyAndCertCmd.PersistentFlags().StringVar(&rootCert, "root-cert", "root-cert.pem", "Output root certificate file")
+	return generateKeyAndCertCmd
+}
+
+func generateCert(name string, client kubernetes.Interface, writer io.Writer) error {
+	names := strings.Split(name, ".")
+	workloadHost := "spiffee://cluster.local/" + names[1] + "/" + names[0]
+	// Signer certificate.
+	var signerCert *x509.Certificate
+	// Signer private key.
+	var signerPriv crypto.PrivateKey
+	var err error
+	var signerCertBytes []byte
+	useK8sAsCA := false
+	dep, err := client.AppsV1().Deployments(istioNamespace).Get("istiod", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	for _, env := range dep.Spec.Template.Spec.Containers[0].Env {
+		if strings.Compare(env.Name, "PILOT_CERT_PROVIDER") == 0 &&
+			strings.Compare(env.Value, "kubernetes") == 0 {
+			useK8sAsCA = true
+			break
+		}
+	}
+	if useK8sAsCA {
+		log.Infof("K8s CA is signing certificate...")
+		k8srootcert, err := saveK8sRootCert(client)
+		if err != nil {
+			return err
+		}
+		certChain, keyPEM, _, err := chiron.GenKeyCertK8sCA(client.CertificatesV1beta1().CertificateSigningRequests(),
+			workloadHost, name+".csr.secret", namespace, k8srootcert)
+		if err != nil {
+			return err
+		}
+		rootCrt, err := ioutil.ReadFile(k8srootcert)
+		if err != nil {
+			return err
+		}
+		err = os.Remove(k8srootcert)
+		if err != nil {
+			return err
+		}
+		saveCreds(certChain, keyPEM, rootCrt, writer)
+	} else {
+		log.Infof("Istiod CA is signing certificate...")
+		caSecret, scrtErr := client.CoreV1().Secrets(istioNamespace).Get(ca.CASecret, metav1.GetOptions{})
+		if scrtErr != nil {
+			return scrtErr
+		}
+		signerCertBytes = caSecret.Data[caCertID]
+		signerPrivBytes := caSecret.Data[caPrivateKeyID]
+		signerCert, err = util.ParsePemEncodedCertificate(signerCertBytes)
+		if err != nil {
+			return fmt.Errorf("pem encoded cert parsing failure (%v)", err)
+		}
+		signerPriv, err = util.ParsePemEncodedKey(signerPrivBytes)
+		if err != nil {
+			return fmt.Errorf("pem encoded key parsing failure (%v)", err)
+		}
+		opts := util.CertOptions{
+			Host:       workloadHost,
+			NotBefore:  getNotBefore(),
+			TTL:        validFor,
+			SignerCert: signerCert,
+			SignerPriv: signerPriv,
+			Org:        org,
+			IsCA:       false,
+			IsClient:   true,
+			RSAKeySize: keySize,
+		}
+		certPem, privPem, err := util.GenCertKeyFromOptions(opts)
+		if err != nil {
+			log.Fatalf("Failed to generate certificate: %v\n", err)
```

Comment:
Why logging and not returning the err?

Reply:
+1, this may be at some point used in istioctl part of preparing the cluster.

It should in fact move to pkg/... - and be used by istioctl to prepare the cluster, in particular for MC cases. I think 'best practice' remains to use intermediary certs - and not copy the same private root in each cluster - and that's why we have an operator for.

Reply:
return error